File: dsp23.pas
(*****  PROGRAM NO.23    DISCRETE FOURIER TRANSFORM (DFT)  *****)
(*******  BY DIRECT CALCULATION, AND WITH TABLE LOOKUP  ********)
(******  REAL SIGNAL, N=64, SPECTRAL MAGNITUDE & PHASE  ********)
program DSP23;
uses graph3;
var X: array[1..64] of real;
var S: array[1..64] of real;
var C: array[1..64] of real;
var XXR: array[1..64] of real;
var XXI: array[1..64] of real;
var MAG: array[1..64] of real;
var PHASE: array[1..64] of real;
var N,K,D,P,J,L,I: integer;
var B,W0,M1,M2,M3: real;
label 1,2;
(**********************  GENERATE SIGNAL  **********************)
begin
    for N:=1 to 64 do begin
      B:=2*3.141593*(N-1);
      X[N]:=sin(B/64)+cos(B/16)+0.6*cos(B/8)+0.5*sin(B/4);end;
  P:=1;
    for N:=1 to 64 do begin XXR[N]:=0;XXI[N]:=0;end;
   writeln('ENTER ANY INTEGER TO START FIRST DFT');readln(D);
     for K:=1 to 64 do begin
      for N:=1 to 64 do begin
       XXR[K]:=XXR[K]+X[N]*cos(2*3.141593*(K-1)*(N-1)/64);
       XXI[K]:=XXI[K]-X[N]*sin(2*3.141593*(K-1)*(N-1)/64);
     end;end;
  if P=1 then goto 2;
1:  for J:=1 to 64 do begin XXR[J]:=0;XXI[J]:=0;end;
    writeln('ENTER ANY INTEGER TO START SECOND DFT');readln(D);
(********************  FORM TABLE LOOKUP  **********************)
W0:=2*3.141593/64;
  for J:=1 to 64 do begin
    C[J]:=cos(W0*(J-1));S[J]:=sin(W0*(J-1));end;
      for K:=1 to 64 do begin
        for N:=1 to 64 do begin
          I:=(K-1)*(N-1);L:=I mod 64;
          XXR[K]:=XXR[K]+X[N]*C[L+1];
          XXI[K]:=XXI[K]-X[N]*S[L+1];
      end;end;
(************  CONVERT TO SPECTRAL MAGNITUDE & PHASE  **********)
2: for K:=1 to 64 do begin
     if abs(XXI[K])<0.001 then XXI[K]:=0;
     if abs(XXR[K])<0.00001 then XXR[K]:=0.00001;
     MAG[K]:=sqrt(XXR[K]*XXR[K] + XXI[K]*XXI[K]);
     PHASE[K]:=arctan(XXI[K]/XXR[K]);end;
(*******  NORMALISE TO MAXIMUM VALUE OF UNITY FOR PLOT  ********)
   M1:=0.001;M2:=0.001;M3:=0.001;
     for N:=1 to 64 do begin
       if abs(X[N])>M1 then M1:=abs(X[N]);
       if abs(MAG[N])>M2 then M2:=abs(MAG[N]);
       if abs(PHASE[N])>M3 then M3:=abs(PHASE[N]);end;
     for N:=1 to 64 do begin
       X[N]:=X[N]/M1;MAG[N]:=MAG[N]/M2;PHASE[N]:=PHASE[N]/M3;end;
(********  PLOT SIGNAL, SPECTRAL MAGNITUDE, AND PHASE  *********)
  graphmode;
    for N:=1 to 64 do begin
      draw (5*(N-1),50,5*(N-1),round(50-X[N]*25),1);
      draw (5*(N-1),120,5*(N-1),round(120-MAG[N]*30),1);
      draw (5*(N-1),160,5*(N-1),round(160-PHASE[N]*20),1);end;
  P:=P+1;if P=2 then goto 1;
end.
(***************************************************************)

File: dsp23.c
#include <graph.h>
#include <conio.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define DIM     64

float   fX[DIM + 1], fXXR[DIM + 1], fXXI[DIM + 1],
        fMAG[DIM + 1], fPHASE[DIM + 1], fC[DIM + 1], fS[DIM + 1];

/*
*** WHEN COMPARING WITH THE TEXT, NOTE THAT IN THE C PROGRAMS **
******* FLOATING-POINT VARIABLES ARE PREFIXED WITH AN f, *******
************** INTEGER VARIABLES WITH AN i *********************
*/

void main (void)
    {
    /*
     ***** PROGRAM NO. 23  DISCRETE FOURIER TRANSFORM (DFT) ****
     ******* BY DIRECT CALCULATION, AND WITH TABLE LOOKUP ******
     ****** REAL SIGNAL, N=64, SPECTRAL MAGNITUDE & PHASE ******
     */
    int    iCount, iCount2, iP, iI, iL;
    float  fMax1, fMax2, fMax3, fB, fW0;

    _clearscreen  (_GCLEARSCREEN);
    printf ("PROGRAM 23: Discrete Fourier Transform (DFT)\n");
    printf ("by direct calculation, and with table lookup.\n");
    printf ("Real signal, N=64, spectral magnitude & phase.\n");
    printf ("Press any key to start first DFT ...");
    getch();
    /*
     ******************** GENERATE SIGNAL **********************
     */
    for (iCount = 1; iCount <= DIM; iCount++)
        {
        fB = 2.0 * 3.141593 * (iCount - 1);
        fX[iCount] = sin (fB / 64.0) + cos (fB / 16.0) +
                     0.6 * cos (fB / 8.0) + 0.5 * sin (fB / 4.0);
        }

    for (iCount = 1; iCount <= DIM; iCount++)
        for (iCount2 = 1; iCount2 <= DIM; iCount2++)
            {
            fXXR[iCount] = fXXR[iCount] + fX[iCount2] *
                          cos (2.0 * 3.141593 * (float)(iCount - 1) *
                          (float)(iCount2 - 1) / 64.0);
            fXXI[iCount] = fXXI[iCount] - fX[iCount2] *
                          sin (2.0 * 3.141593 * (float)(iCount - 1) *
                          (float)(iCount2 - 1) / 64.0);
            }

    for (iP = 0; iP < 2; iP++)
        {
        _setvideomode( _MRES16COLOR );

        if (iP == 0)
           printf ("\n\nPress any key to start second DFT...");

        if (iP)
            {
            for (iCount = 1; iCount <= DIM; iCount++)
                {
                fXXR[iCount] = 0.0;
                fXXI[iCount] = 0.0;
                }

            /******** FORM TABLE LOOKUP ***********************/

            fW0 = 2.0 * 3.141593 / 64.0;

            for (iCount = 1; iCount <= DIM; iCount++)
                {
                fC[iCount] = cos (fW0 * ((float) (iCount - 1)));
                fS[iCount] = sin (fW0 * ((float) (iCount - 1)));
                }

            for (iCount = 1; iCount <= DIM; iCount++)
                for (iCount2 = 1; iCount2 <= DIM; iCount2++)
                    {
                    iI = (iCount -1) * (iCount2 - 1);
                    iL = iI % 64;

                    fXXR[iCount] = fXXR[iCount] + fX[iCount2] *
                        fC[iL + 1];
                    fXXI[iCount] = fXXI[iCount] - fX[iCount2] *
                        fS[iL + 1];
                    }

            }

        /*
         ********* CONVERT TO SPECTRAL MAGNITUDE AND PHASE *****
         */
        for (iCount = 1; iCount <= DIM; iCount++)
            {
            if (fabs (fXXI[iCount]) < 0.001)
                fXXI[iCount] = 0.0;
            if (fabs (fXXR[iCount]) < 0.00001)
                fXXR[iCount] = 0.00001;
            fMAG[iCount] = sqrt (pow (fXXR[iCount], 2.0) +
                            pow (fXXI[iCount], 2.0));
            fPHASE[iCount] = atan (fXXI[iCount] / fXXR[iCount]);
            }

        /*
         ******* NORMALISE TO MAXIMUM OF UNITY FOR PLOT ********
         */
        fMax1 = fMax2 = fMax3 = 0.001;

        for (iCount = 1; iCount <= DIM; iCount++)
            {
            if (fabs (fX[iCount]) > fMax1)
                fMax1 = fabs (fX[iCount]);
            if (fabs (fMAG[iCount]) > fMax2)
                 fMax2 = fabs (fMAG[iCount]);
            if (fabs (fPHASE[iCount]) > fMax3)
                fMax3 = fabs (fPHASE[iCount]);
            }

        for (iCount = 1; iCount <= DIM; iCount++)
            {
            fX[iCount] /= fMax1;
            fMAG[iCount] /= fMax2;
            fPHASE[iCount] /= fMax3;
            }

        /*
         ********PLOT SIGNAL, SPECTRAL MAGNITUDE AND PHASE *****
         */
        for (iCount = 1; iCount <= DIM; iCount++)
            {
            _moveto (5 * (iCount - 1), 50);
            _lineto (5 * (iCount - 1), (int)(50 - fX[iCount] *
                25.0 + 0.5));
            _moveto (5 * (iCount - 1), 120);
            _lineto (5 * (iCount - 1), (int)(120 - fMAG[iCount] *
                30.0 + 0.5));
            _moveto (5 * (iCount - 1), 160);
            _lineto (5 * (iCount - 1), (int)(160 - fPHASE[iCount] *
                20.0 + 0.5));
            }

        getch();
        _setvideomode( _DEFAULTMODE );
        }

    /*
     ***********************************************************
     */
    }
